### PR TITLE
Enable dynamic cloud volume modifications

### PR DESCRIPTION
**Pull request by gberginc**
Recently Amazon introduced a capability to modify existing cloud volumes
on the fly. Depending on the volume type size, volume type and IOPS can
be modified without detaching the volume first.

This patch enhances the existing `update_volume` method that previously
allowed changing only the name of the volume by using the new method
`#modify_volume` from the aws-sdk gem. The patch also adds a set of
tests verifying different possible options passed to the
`update_volume`.

Depends on #176 (requires newer aws-sdk gem).

@miq-bot add_label enhancement

### CODE REVIEW

File: manageiq-providers-amazon.gemspec [hunk 1]
```diff
@@ -13,7 +13,7 @@ Gem::Specification.new do |s|
 
   s.files = Dir["{app,config.lib}/**/*"]
 
-  s.add_dependency("aws-sdk", ["~>2.6.14"])
+  s.add_dependency("aws-sdk", ["~>2.7.8"])
```
Comment (durandom) [inline manageiq-providers-amazon.gemspec]: can you put this in a separate PR - for backporting reasons...

Comment (durandom) [inline manageiq-providers-amazon.gemspec]: oh, it's already in #176  👍

**Comment by gberginc** [comment]
Since #176 was just merged, I rebased it and removed the gem version bump from this one.

File: spec/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume_spec.rb [hunk 2]
```diff
@@ -35,6 +35,139 @@
...
     end
 
+    context "#update_volume" do
+      let(:modify_volume_stub) { { :modify_volume => {} } }
+
+      context "updating volume name" do
+        it "is allowed when :name is provided" do
+          # Name is set via AWS tags so stub the create_tags.
+          expect_any_instance_of(Aws::EC2::Volume).to receive(:create_tags).with(
+            :tags => [{:key => "Name", :value => "new_name"}]
```
Comment (durandom) [inline spec/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume_spec.rb]: I think you can also stub this. add the api calls to `with_aws_stubbed` below

File: app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb [hunk 3]
```diff
@@ -33,12 +33,17 @@ def validate_update_volume
   end
 
   def raw_update_volume(options)
-    # Only name update is currently supported
-    if options.key?(:name)
-      with_provider_object do |volume|
-        volume.create_tags(:tags => [{:key => "Name", :value => options[:name]}])
-      end
+    res = nil
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb]: does any caller really care about the return value?

File: app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb [hunk 4]
```diff
@@ -92,4 +97,17 @@ def provider_object(connection = nil)
...
     connection.volume(ems_ref)
   end
+
+  private
+
+  def modify_volume_options(options = {})
+    modify_opts = {}
+    if volume_type != 'standard'
+      modify_opts[:volume_type] = options[:volume_type] if options[:volume_type] && options[:volume_type] != 'standard'
+      modify_opts[:size]        = options[:size] if options[:size] && options[:size] != (size / 1_073_741_824)
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb]: Is this something like `1.gigabyte` ?

File: spec/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume_spec.rb (hunk 2 above)
Comment (gberginc) [inline spec/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume_spec.rb]: Possible, but only to some degree. I wanted to ensure that specific arguments are used as well as check whether the call is made or not (if name is given, it must be called, if it's not given, it should not be called).

What I could do is ignore input parameters to `create_tags` and then use an exception in the stub to ensure it doesn't get called. I'll have a look.

File: app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb (hunk 3 above)
Comment (gberginc) [inline app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb]: I am not completely sure. Perhaps I just misunderstood what's needed.

File: app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb (hunk 4 above)
Comment (gberginc) [inline app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb]: true, alternative check would be `... && options[:size].gigabyte == size`. Database stores the size in bytes, but the `:size` in options must be in gigabytes.

**Comment by gberginc** [comment]
@durandom I've updated the change:
 * use stubs for all tests: in these stubs I have explicitly set exceptions for the calls for which I'd like to ensure they are not called (for example, when `:name` is not given, it should not call `create_tags`, so I set the stubbed response to `failed` to ensure exception is thrown in case the call is made)
 * use `gigabyte` instead of that ugly number
 * ignore return value of `update_volume` - it really doesn't seem like it's used somewhere

File: app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb [hunk 5]
```diff
@@ -92,4 +94,17 @@ def provider_object(connection = nil)
...
     connection.volume(ems_ref)
   end
+
+  private
+
+  def modify_volume_options(options = {})
+    modify_opts = {}
+    if volume_type != 'standard'
+      modify_opts[:volume_type] = options[:volume_type] if options[:volume_type] && options[:volume_type] != 'standard'
+      modify_opts[:size]        = options[:size] if options[:size] && options[:size].gigabytes != size
```
Comment (durandom) [inline app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb]: where does `options[:size]` come from? Maybe do `Integer(options[:size])` ?

Comment (gberginc) [inline app/models/manageiq/providers/amazon/storage_manager/ebs/cloud_volume.rb]: @durandom Thanks for spotting this. So far value of `options[:size]` came from the UI where I tested only with valid values, however it could also come from API or automate which means that it could in fact be invalid. I added two more tests to ensure that valid and invalid strings work properly with the updated function.

**Comment by Ladas** [comment]
looks great, very nice specs :+1:
